plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -311,38 +315,118 @@ costIsAcceptable = \case
   TyInst{}   -> False
   Let{}      -> False
 
--- See Note [Inlining criteria]
--- | Is the size increase (in the AST) of inlining a variable whose RHS is
--- the given term acceptable?
-sizeIsAcceptable :: Term tyname name uni fun ann -> Bool
-sizeIsAcceptable = \case
-  Builtin{}  -> True
-  Var{}      -> True
-  Error{}    -> True
-  LamAbs {}  -> False
-  TyAbs {}   -> False
+-- State type used by `sizeIsAcceptable`.
+-- See Note [Size criteria for inlining].
+data St a = St
+    { _stRemainingSize :: Int
+    , _stUnseenParams  :: Set a
+    -- ^ Parameter names in the LHS that haven't appeared in the RHS.
+    , _stSeenParams    :: Set a
+    -- ^ Parameter names in the LHS that have appeared in the RHS.
+    }
 
-  -- Inlining constructors of size 1 or 0 seems okay
-  Constr _ _ _ es  -> case es of
-      []  -> True
-      [e] -> sizeIsAcceptable e
-      _   -> False
-  -- Cases are pretty big, due to the case branches
-  Case{} -> False
-
-  -- Arguably we could allow these two, but they're uncommon anyway
-  IWrap{}    -> False
-  Unwrap{}   -> False
-  -- Constants can be big! We could check the size here and inline if they're
-  -- small, but probably not worth it
-  Constant{} -> False
-  Apply{}    -> False
-  TyInst{}   -> False
-  Let{}      -> False
+makeLenses ''St
+
+-- See Note [Inlining criteria] and Note [Size criteria for inlining].
+-- | Is the size increase (in the AST) of the given inlining acceptable?
+sizeIsAcceptable ::
+    forall name tyname uni fun ann.
+    Ord name =>
+    -- | The set of term parameter names in the definition of the variable.
+    -- e.g., for @let x = f y@, the set is empty;
+    -- for @let x = \a. /\b. c. f y@, the set is @[a, c]@.
+    Set name ->
+    -- | The RHS whose size we are evaluating
+    Term tyname name uni fun ann ->
+    Bool
+sizeIsAcceptable params =
+    flip
+        evalState
+        ( -- See Note [Size criteria for inlining].
+          St
+            { _stRemainingSize = 1 + 2 * length params
+            , _stUnseenParams = params
+            , _stSeenParams = mempty
+            }
+        )
+        . go
+  where
+    go :: Term tyname name uni fun ann -> State (St name) Bool
+    go term = ifM (use stRemainingSize <&> (< 0)) (pure False) $ do
```

Comment:
I think we should use a standalone size metric, for example the one defined in `Size.hs`. Then we can just compare it, rather than this complicated mix of computing it and traversing.